sky/backends/backend_utils.py

```diff
@@ -1985,19 +1988,61 @@ def refresh_cluster_record(
         return None
     check_owner_identity(cluster_name)
 
-    handle = record['handle']
-    if isinstance(handle, backends.CloudVmRayResourceHandle):
-        use_spot = handle.launched_resources.use_spot
-        has_autostop = (record['status'] != status_lib.ClusterStatus.STOPPED and
-                        record['autostop'] >= 0)
-        force_refresh_for_cluster = (force_refresh_statuses is not None and
-                                     record['status'] in force_refresh_statuses)
-        if force_refresh_for_cluster or has_autostop or use_spot:
-            record = _update_cluster_status(
-                cluster_name,
-                acquire_per_cluster_status_lock=acquire_per_cluster_status_lock,
-                cluster_status_lock_timeout=cluster_status_lock_timeout)
-    return record
+    if not isinstance(record['handle'], backends.CloudVmRayResourceHandle):
+        return record
+
+    # The loop logic allows us to notice if the status was updated in the
+    # global_user_state by another process and stop trying to get the lock.
+    # The core loop logic is adapted from FileLock's implementation.
+    start_time = time.perf_counter()
+    lock = filelock.FileLock(CLUSTER_STATUS_LOCK_PATH.format(cluster_name))
+
+    try:
+        # Loop until we have an up-to-date status or until we acquire the lock.
+        while True:
+            # Check to see if we can return the cached status.
+            if not _must_refresh_cluster_status(record, force_refresh_statuses):
+                return record
+
+            if not acquire_per_cluster_status_lock:
+                return _update_cluster_status_no_lock(cluster_name)
+
+            # Try to acquire the lock so we can fetch the status.
+            try:
+                lock.acquire(blocking=False)
+                # Lock acquired.
```

Comment:
Use context manager to do locking and minimize lock handling in our code?

Reply:
Thanks. Got it in my head that this wouldn't work with the loop logic but that's not right.

Reply:
Fun fact about FileLock, it also implements `__del__` so technically it should get released whene it goes out of scope. (Not that we should depend on that.!